File: src/main/leetcode/p307/NumArray.java
package p307;

/**
 * Created by wangsenyuan on 26/10/2016.
 */
public class NumArray {
    private long[] sums;
    private int[] nums;

    public NumArray(int[] nums) {
        this.nums = nums;
        int n = nums.length;
        this.sums = new long[n + 1];

        for (int i = 0; i < n; i++) {
            doUpdate(i, nums[i]);
        }
    }

    void doUpdate(int i, int delta) {
        int index = i + 1;
        while (index < sums.length) {
            this.sums[index] += delta;
            index += lsb(index);
        }
    }

    void update(int i, int val) {
        doUpdate(i, val - nums[i]);
        nums[i] = val;
    }

    public int sumRange(int i, int j) {
        return getSum(j) - getSum(i - 1);
    }


    private int getSum(int i) {
        int index = i + 1;
        int sum = 0;
        while (index > 0) {
            sum += this.sums[index];
            index -= lsb(index);
        }
        return sum;
    }

    private int lsb(int i) {
        return i & (-i);
    }

    public static void main(String[] args) {
        int[] nums = {1, 3, 5};
        NumArray numArray = new NumArray(nums);
        System.out.println(numArray.sumRange(0, 2));
        numArray.update(1, 2);
        System.out.println(numArray.sumRange(0, 2));
    }
}

File: src/main/leetcode/set0000/set300/set300/p307/NumArray.scala
package set0000.set300.set300.p307

class NumArray(_nums: Array[Int]) {

  val n = _nums.length

  val bit = Array.ofDim[Int](n + 1)

  private def add(i: Int, v: Int): Unit = {
    var j = i + 1
    while (j <= n) {
      bit(j) += v
      j += j & (-j)
    }
  }

  (0 until n) foreach (i => add(i, _nums(i)))

  def update(i: Int, `val`: Int) {
    val diff = `val` - _nums(i)
    add(i, diff)
    _nums(i) = `val`
  }

  private def sum(i: Int): Int = {
    var j = i + 1
    var res = 0
    while (j > 0) {
      res += bit(j)
      j -= j & (-j)
    }
    res
  }

  def sumRange(i: Int, j: Int): Int = {
    sum(j) - sum(i - 1)
  }

}
